08_Loading_Elf/02_Elf_Theory.md

```diff
@@ -0,0 +1,117 @@
+
+This is actually how the `.bss` section is allocated, and any pre-zeroed parts of an ELF executable are created this way. 
+
+Before looking at some example code your VMM will need a new function that tries to allocate memory at a *specific* virtual address, instead of whatever is the best fit. For our example we're going to assume you have the following implemented, but adjust to your own design:
+
+```c
+void* vmm_alloc_at(uintptr_t addr, size_t length, size_t flags);
+```
+
+Alternatively you could make use of the extra argument in `vmm_alloc`, and add a new flag like `VM_FLAG_AT_ADDR` that indicates the VMM should use the extra arg as the virtual address.
+
+The reason we need to use a specific address is that the code and data contained in the ELF are compiled and linked assuming that they're at that address. There might be code that jumps to a fixed address or data that is expected to be at a certain address. If we don't copy the program header where it expects to be, the program may break.
+
+*Authors Note: Relocations are another way of dealing with the problem of not being to use the requested virtual address, but these are more advanced. They're not hard to implement, certainly easier than dynamic linking, but still beyond the scope of this section.*
+
+Now that we have that, lets look at the example code (without error handling, as always):
+
+```c
+void load_phdr(Elf64_EHdr* ehdr, Elf64_Phdr* phdr) {
+    if (phdr->p_type != PT_LOAD)
+        return;
+    
+    void* dest = vmm_alloc_at(phdr->p_vaddr, phdr->p_memsz, VM_FLAG_WRITE);
+    memcpy(dest, (void*)ehdr + phdr->p_offset, phdr->p_filesz);
+
+    const zero_count = phdr->p_memsz - phdr->p_filesz;
+    memset(dest + phdr->p_filesz, 0, zero_count);
+}
+```
+
+### Program Header Flags
+
+At this point we've got the program header's content loaded in the correct place. We'll run into an issue if we try to use the loaded program header in this state: we've mapped all program headers as read/write/no-execute. This means if we try to execute any of the headers as code (and at least one of them is guarenteed to be code), we'll fault.
+
+Fortunately the solution is quite straight forward, the read/write/execute permissions required for a phdr are encoded in the `p_flags` field. This field is actually a bitfield, with bit 0 representing execute, bit 1 representing write and bit 2 representing read. You can ignore the read permission as most platform's dont allow for write-only memory anyway, but you'll want to be sensitive to the write and execute permissions.
```

Comment:
straightforward can be also written as 2 separate words? or is it a typo?

Reply:
platform's ?

Reply:
should we use zero, one, two instead of 0 1 and 2? since we are in the middle of the text, it should be better (or maybe not)?